File: src/main/scala/booklet/impure/service/BookHandler.scala
package booklet.impure.service

import booklet.pure.Failure
import booklet.pure.http.CustomResponse._
import booklet.pure.http.Query
import booklet.pure.model.{BookData, BookId}
import booklet.pure.views.BookView
import zhttp.http.Path.Segment
import zhttp.http.{Body, Path, Request, Response}
import zio.{IO, UIO, URLayer, ZIO, ZLayer}

trait BookHandler {
  def fetchAll(userId: String): IO[Failure, Response]

  def fetch(bookId: String, userId: String): IO[Failure, Response]

  def create(request: Request, userId: String): IO[Failure, Response]

  def update(bookId: String, request: Request, userId: String): IO[Failure, Response]

  def delete(bookId: String, userId: String): IO[Failure, Response]
}

object BookHandler {
  def fetchAll(userId: String): ZIO[BookHandler, Failure, Response] =
    ZIO.serviceWithZIO(_.fetchAll(userId))

  def fetch(bookId: String, userId: String): ZIO[BookHandler, Failure, Response] =
    ZIO.serviceWithZIO(_.fetch(bookId, userId))

  def create(request: Request, userId: String): ZIO[BookHandler, Failure, Response] =
    ZIO.serviceWithZIO(_.create(request, userId))

  def update(
      bookId: String,
      request: Request,
      userId: String
  ): ZIO[BookHandler, Failure, Response] =
    ZIO.serviceWithZIO(_.update(bookId, request, userId))

  def delete(bookId: String, userId: String): ZIO[BookHandler, Failure, Response] =
    ZIO.serviceWithZIO(_.delete(bookId, userId))
}

object BookHandlerLive {

  private val booksPath = "/books"

  private def toBookId(bookId: String) =
    ZIO
      .fromOption(bookId.toLongOption)
      .map(BookId(_))

  private def fetchAllImpl(db: GoogleSheetsService, userId: String) =
    db.fetchAllBooks(userId)
      .fold(
        serverFailure,
        books => ok(Body.fromString(BookView.list(books)))
      )

  private def fetchImpl(db: GoogleSheetsService, bookId: String, userId: String) =
    toBookId(bookId)
      .foldZIO(
        _ => ZIO.succeed(badRequest(s"Cannot parse ID $bookId")),
        id =>
          db
            .fetchBook(id, userId)
            .fold(
              serverFailure,
              {
                case None       => notFound(Path(Vector(Segment(bookId))))
                case Some(book) => ok(Body.fromString(BookView.list(Seq(book))))
              }
            )
      )

  private def createImpl(db: GoogleSheetsService, request: Request, userId: String) =
    Query
      .fromRequest(request)
      .flatMap(requestQry =>
        ZIO
          .fromOption(BookData.completeFromHttpQuery(requestQry))
          .foldZIO(
            _ => ZIO.succeed(badRequest(requestQry.toString)),
            bookData =>
              db
                .insertBook(bookData.copy(userId = Some(userId)), userId)
                .fold(
                  serverFailure,
                  _ => seeOther(booksPath)
                )
          )
      )

  private def updateImpl(
      db: GoogleSheetsService,
      bookId: String,
      request: Request,
      userId: String
  ) =
    Query
      .fromRequest(request)
      .flatMap(requestQry =>
        toBookId(bookId)
          .foldZIO(
            _ => ZIO.succeed(badRequest(requestQry.toString)),
            id =>
              db
                .updateBook(id, BookData.partialFromHttpQuery(requestQry), userId)
                .fold(
                  serverFailure,
                  _ => seeOther(booksPath)
                )
          )
      )

  private def deleteImpl(db: GoogleSheetsService, bookId: String, userId: String) =
    toBookId(bookId)
      .foldZIO(
        _ => ZIO.succeed(badRequest(s"Cannot parse ID $bookId")),
        id =>
          db
            .deleteBook(id, userId)
            .fold(
              serverFailure,
              _ => seeOther(booksPath)
            )
      )

  private def fromDatabase(db: GoogleSheetsService) =
    new BookHandler {
      override def fetchAll(userId: String): UIO[Response] = fetchAllImpl(db, userId)

      override def fetch(bookId: String, userId: String): UIO[Response] =
        fetchImpl(db, bookId, userId)

      override def create(request: Request, userId: String): IO[Failure, Response] =
        createImpl(db, request, userId)

      override def update(bookId: String, request: Request, userId: String): IO[Failure, Response] =
        updateImpl(db, bookId, request, userId)

      override def delete(bookId: String, userId: String): UIO[Response] =
        deleteImpl(db, bookId, userId)
    }

  val layer: URLayer[GoogleSheetsService, BookHandler] =
    ZLayer.fromZIO(ZIO.service[GoogleSheetsService].map(fromDatabase))
}

File: src/main/ts/booklet/impure/service/BookHandler.ts
import { Failure } from "../../pure/Failure";
import { BookData, BookId } from "../../pure/model/BookData";
import { GoogleSheetsService } from "./GoogleSheetsService";
import { BookView } from "../../pure/views/BookView";
import { Query } from "../../pure/http/Query";
import { CustomResponse } from "../../pure/http/CustomResponse";
import { Request, Response } from "express";

export interface BookHandler {
  fetchAll(userId: string): Promise<Response>;
  fetch(bookId: string, userId: string): Promise<Response>;
  create(request: Request, userId: string): Promise<Response>;
  update(bookId: string, request: Request, userId: string): Promise<Response>;
  delete(bookId: string, userId: string): Promise<Response>;
}

export class BookHandlerLive implements BookHandler {
  private googleSheetsService: GoogleSheetsService;

  constructor(googleSheetsService: GoogleSheetsService) {
    this.googleSheetsService = googleSheetsService;
  }

  async fetchAll(userId: string): Promise<Response> {
    try {
      const books = await this.googleSheetsService.fetchAllBooks(userId);
      return CustomResponse.ok(BookView.list(books));
    } catch (error) {
      return CustomResponse.serverFailure(Failure.fromThrowable(error));
    }
  }

  async fetch(bookId: string, userId: string): Promise<Response> {
    try {
      const id = this.toBookId(bookId);
      const book = await this.googleSheetsService.fetchBook(id, userId);
      if (book) {
        return CustomResponse.ok(BookView.list([book]));
      } else {
        return CustomResponse.notFound(bookId);
      }
    } catch (error) {
      return CustomResponse.serverFailure(Failure.fromThrowable(error));
    }
  }

  async create(request: Request, userId: string): Promise<Response> {
    try {
      const requestQry = Query.fromRequest(request);
      const bookData = BookData.completeFromHttpQuery(requestQry);
      if (bookData) {
        await this.googleSheetsService.insertBook(userId, bookData);
        return CustomResponse.seeOther("/books");
      } else {
        return CustomResponse.badRequest(requestQry.toString());
      }
    } catch (error) {
      return CustomResponse.serverFailure(Failure.fromThrowable(error));
    }
  }

  async update(bookId: string, request: Request, userId: string): Promise<Response> {
    try {
      const requestQry = Query.fromRequest(request);
      const id = this.toBookId(bookId);
      const bookData = BookData.partialFromHttpQuery(requestQry);
      await this.googleSheetsService.updateBook(userId, id, bookData);
      return CustomResponse.seeOther("/books");
    } catch (error) {
      return CustomResponse.serverFailure(Failure.fromThrowable(error));
    }
  }

  async delete(bookId: string, userId: string): Promise<Response> {
    try {
      const id = this.toBookId(bookId);
      await this.googleSheetsService.deleteBook(userId, id);
      return CustomResponse.seeOther("/books");
    } catch (error) {
      return CustomResponse.serverFailure(Failure.fromThrowable(error));
    }
  }

  private toBookId(bookId: string): BookId {
    const id = parseInt(bookId, 10);
    if (isNaN(id)) {
      throw new Error(`Cannot parse ID ${bookId}`);
    }
    return new BookId(id);
  }
}
